---
title: "Taxonomic Assignment"
author: "Kimberly Ledger"
date: "2024-05-03"
output: html_document
---

latest updates - 7 August 2024

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load-libraries}
library(tidyverse)
library(tidyr)
library(stringr)
rename <- dplyr::rename
```

## load output from taxonkit 
- this is the fasta seqs from the 20240806 run on dadasnake (i.e. duplicate sbs2022 libraries separated), and blastn using a ncbi nt database accessed on 2024-07-18 with 98% seq identity and 98% query coverage   
```{r load-output-from-taxonkit}
taxonomy <-read.delim("/home/kimberly.ledger/dbo_metabarcoding/data/NBS_SBS_DBO_20240806/blastn_taxlineage_98.txt", header = FALSE, na.strings=c(""," ","NA"))

head(taxonomy)
```


## clean up the header a bit
```{r}
# use the full taxonomy rather than the seq id to collapse identical entries
tax_df <- taxonomy %>%
  filter(V4 > 100) %>% # make sure all retained matches are >100 bp
  select(-V2, -V5, -V6, -V7, -V14, -V8, -V9, -V10, -V11, -V12, -V13, -V15, -V16) %>%  #remove unnecessary columns
  group_by(V1, V17) %>% # group by the sequence key and the full taxonomy to reduce duplicate entries
  unique() %>% # doing that reduced the number of entries
  rename(qseqid=V1, perc_id=V3, length=V4, taxonomy=V17) %>% #rename headers
  filter(perc_id > 98) %>% 
  filter(!is.na(taxonomy))
```

## calcualate the number of ASV's with taxonomic assignment
```{r how-many-total-ASVs-w-taxonomy}
tax_df %>% 
  ungroup() %>%
  select(qseqid) %>%
  unique() %>%
  nrow
```

## reformat table 
```{r}
# formatting the taxonomy variables
taxon_df <- tax_df %>%
  filter(str_detect(taxonomy, ";")) %>%
  separate(taxonomy, into=c("kingdom", "phylum", "class", "order", "family", "genus", "species"), sep = ";") %>% 
  filter(!grepl("sp\\.", species, ignore.case = TRUE)) %>%
  filter(!grepl("cf\\.", species, ignore.case = TRUE)) %>%
  filter(!grepl("aff\\.", species, ignore.case = TRUE)) %>%
  filter(!grepl(" x ", species, ignore.case = TRUE)) %>%         ### remove hybrids and unknown species assignments
  filter(!grepl("\\d", species))
```


## figure out what ASVs are unambiguously a single species
```{r}
single_spp_seqs <- taxon_df %>% 
  group_by(qseqid) %>%
  add_tally(name = "n_taxa") %>%
  filter(n_taxa == 1)
```

## remove the single-species seqs from the dataframe and then rank the hits by % identity for the remaining seqs
```{r}
seq_id_diff <- taxon_df %>%
  anti_join(., single_spp_seqs) %>%  ## comment out this line if there are no single species ASVs
  select(-length) %>%
  group_by(qseqid, species, genus, family, order, class, phylum, kingdom) %>%
    mutate(seq_percID = max(perc_id)) %>%
    group_by(qseqid, species, genus, family, order, class, phylum, kingdom, seq_percID) %>%
  summarise(max(seq_percID)) %>% # take just the top hit for each taxon (for each sequence)
  select(-`max(seq_percID)`) %>%
  ungroup() %>%
  group_by(qseqid) %>%
      #mutate(id_rank = rank(desc(seq_percID), ties.method = "min")) %>% # rank the taxonomic hits per sequence by % id
       mutate(top_perc = max(seq_percID)) %>% # designate the highest % id for the best taxonomic hit in each sequence (in some, but not all cases, this is 100%)   %>%
      mutate(diff = top_perc - seq_percID) %>% # calculate the difference between the % identity of the top hit and each subsequent taxonomic hit
        arrange(diff)
```

Now you have the single best entry for each species for each sequence ranked and with the difference between the first and second ranked entries calculated.

For sequences with multiple top hits, where the difference between ranked taxa = 0, you will end up defaulting to genus- or family-level ID. We'll do the same for any sequences where the difference between the first and second ranked taxa is < 1%.

** if there is a sequence that is a 100% match, then keep just the sequences within 0.5%  
** if there is NOT a 100% match, keep the top 1% 

## ASVs with 100% match 
```{r}
match_100 <- seq_id_diff %>%
  filter(top_perc == 100) %>%
  select(qseqid) %>%
  unique()

match_lessthan100 <- seq_id_diff %>%
  filter(top_perc != 100) %>%
  select(qseqid) %>%
  unique()
```


## Figure out which differences are **> 1%** and eliminate those first
```{r}
# filter out any taxa that are >1% less matching identity than the top taxonomic hit for a given sequence
match_100_to_remove_low_perc_hits <- seq_id_diff %>%
  ungroup() %>%
  filter(qseqid %in% match_100$qseqid) %>%
  group_by(qseqid) %>%
  filter(diff > 0.5)

match_lessthan100_to_remove_low_perc_hits <- seq_id_diff %>%
  ungroup() %>%
  filter(qseqid %in% match_lessthan100$qseqid) %>%
  group_by(qseqid) %>%
  filter(diff > 1)

to_remove_low_perc_hits <- match_100_to_remove_low_perc_hits %>%
  bind_rows(match_lessthan100_to_remove_low_perc_hits)

keepers <- seq_id_diff %>%
  anti_join(to_remove_low_perc_hits)
```

## so now determine whether ASVs should be assigned to genus, family, order, etc. 
```{r}
singletons <- keepers %>%
  select(qseqid) %>%
  tally() %>%
  filter(n == 1)

# these are the seqs that now have only a single match
singleton_df <- singletons %>%
  left_join(keepers) %>%
  select(-n) %>%
  bind_rows(single_spp_seqs) %>% # combine the single spp data
  mutate(taxonomic_level = "species") %>%
  mutate(taxon = species)

## Genus-level matches
# remove the singletons from the bigger df 
single_genus <- keepers %>%
  anti_join(singleton_df)%>% # at best, these should be genus-level matches
  group_by(qseqid, genus) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single genus
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "genus") %>%
  mutate(taxon = genus)
  

## Family-level matches
single_family <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  group_by(qseqid, family) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single family
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "family") %>%
  mutate(taxon = family)


## Order-level matches
single_order <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  group_by(qseqid, order) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single order
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "order") %>%
  mutate(taxon = order)


## Class-level matches
single_class <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  group_by(qseqid, class) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single class
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "class") %>%
  mutate(taxon = class)


## Phylum-level matches
single_phylum <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  anti_join(single_class) %>%
  group_by(qseqid, phylum) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single phylum
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "phylum") %>%
  mutate(taxon = phylum)

# Kingdom level matches
single_kingdom <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  anti_join(single_class) %>%
  group_by(qseqid, kingdom) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single kingdom
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "kingdom") %>%
  mutate(taxon = kingdom)
```

## modify the singleton_df to include the right variable headers
```{r}
single_spp <- singleton_df %>%
  select(-perc_id, -length, -n_taxa) %>%
  mutate(taxonomic_level = "species") %>%
  mutate(taxon = species)
```

## recombine the full data set now that the appropriate level of taxonomy has been determined
```{r}
tax_df_2 <- bind_rows(single_kingdom, single_phylum, single_class, single_order, single_family, single_genus, single_spp)
```

# now let's take a closer look at the assignments we are getting from inital blastn 

## what non-fish are here? 
```{r}
not_Actinopteri <- tax_df_2 %>%
  filter(class != "Actinopteri") %>%
  select(species, genus, family, order, class, phylum) %>%
  unique()
```

## remove terrestrial/freshwater/out-of-range non-fish from data set 
```{r}
not_Actinopteri_keepers <- not_Actinopteri %>% 
  #class = Asteroidea - species = Asterias amurensis (North Pacific seastar) - seem reasonable to keep
  #class = Aves - keeping murre and murrelet
  filter(order != "Galliformes") %>%  ### remove chicken, junglefowl 
  filter(species != "Harpagornis moorei") %>% ## definitely no extinct eagle from New Zealand... 
  #family = Petromyzontidae - lamprey
  filter(!(family == "Petromyzontidae" & species != "Lethenteron camtschaticum")) %>% ### remove lamprey except Arctic lamprey 
  #picoplankton
  filter(species != "Bathycoccus prasinos") %>% ### remove a picoplankton
  #class = mammals
  #order = Artiodactyla
  filter(!family == "Bovidae") %>% ### remove bovids
  filter(!(family == "Delphinidae" & species != "Lagenorhynchus obliquidens")) %>% ### remove dolphins expect Pacific white-sided 
  filter(!family == "Suidae") %>% ### remove pigs 
  filter(!family == "Cervidae") %>% ### remove moose 
  #order = Carnivora
  filter(!species == "Pusa sibirica") %>% ### remove Baikal seal 
  filter(!species == "Pusa caspica") %>% ### remove Caspian seal 
  filter(!family == "Hominidae") %>% ### remove humans
  filter(!family == "Canidae") %>% ### remove dog
  select(species) %>%
  rename(Species = species) %>%
  mutate(in_range = "yes")
```

## now look at the fish and figure out what taxa are in/out of our range
```{r}
to_check_range <- tax_df_2 %>%
  filter(class == "Actinopteri") %>%
  select(species, genus, family, order, class, phylum) %>%
  unique()
```

## check ranges for species using rfishbase 
## also at this step, check 
```{r}
#library(remotes)
#remotes::install_github("ropensci/rfishbase")
library(rfishbase)

#one-time download of all fishbase tables... this takes a bit 
fb_import()

#first, validate species names using rfishbase synonyms
spp_df <- synonyms(to_check_range$species)

syn <- spp_df %>% 
  filter(Status == "synonym")
  
to_check_range <- to_check_range %>% 
  mutate(validated_name = ifelse(species %in% syn$synonym, syn$Species, species))
    
to_check_range %>%
  filter(species != validated_name)

#get distribution info 
spp_distribution <- distribution(to_check_range$validated_name) %>%
  select(Species, FAO) %>%
  unique()

#add column to designate if we will consider a species as "in range"- for this study, this will be NE Pacific and Arctic Ocean 
spp_distribution_range <- spp_distribution %>%
  mutate(in_range = ifelse(is.na(FAO), NA, "no"),
         in_range = ifelse(FAO == "Pacific, Northeast", "yes", in_range),
         in_range = ifelse(FAO == "Arctic Ocean", "yes", in_range))

#keep just a list of spp names and yes/no/NA for "in range"  - this way we can keep track of what spp didn't have any reference information in fishbase to determine range 
spp_range <- spp_distribution_range %>%
  select(Species, in_range) %>%
  unique()

#how many entries do not have range info? 
range_na <- spp_range %>%
  filter(is.na(in_range))
```

When a valid name was not found, the presence of a species in the study area was checked using the GBIF database (https://www.gbif.org/).

# some species do not have range info - manually determine if these species should be considered in range
```{r}
range_na <- range_na %>%
  mutate(in_range = ifelse(Species == "Ammodytes japonicus", "no", in_range),
         #in_range = ifelse(Species == "Cleisthenes herzensteini", "no", in_range),
         #in_range = ifelse(Species == "Gadus ogac", "no", in_range),
         #in_range = ifelse(Species == "Myoxocephalus aenaeas", "no", in_range),
         #in_range = ifelse(Species == "Kareius bicoloratus", "no", in_range),
         #in_range = ifelse(Species == "Cottocomephorus grewingki", "no", in_range),
         in_range = ifelse(Species == "Sebastes cheni", "no", in_range)) #,
         #in_range = ifelse(Species == "Embassichthys bathybius", "yes", in_range),    ##new name is Microstomus bathybius
         #in_range = ifelse(Species == "Pungitius kaibarae", "no", in_range),
         #in_range = ifelse(Species == "Ulcina olrikii", "yes", in_range),  ### arctic alligatorfish - Aspidophoroides olrikii
         #in_range = ifelse(Species == "Polypera greeni", "yes", in_range))  ## Liparis greeni
```


- in addition to fishbase distributions, these citations may help: https://www.marinespecies.org/arms/aphia.php?p=taxdetails%26id=10194
and exclude Atlantic Arctic species: Mecklenburg, C. W., Lynghammar, A., Johannesen, E., Byrkjedal, I., Christiansen, J. S., Dolgov, A. V., Karamushko, O. V., Mecklenburg, T. A., Møller, P. R., Steinke, D., & Wienerroither, R. M. (2018). Marine fishes of the Arctic region. Conservation of Arctic Flora and Fauna.

## manually change a few species that are not really in the range.. and combine dataframes

--- come back to this step and remove more Atlantic Arctic species. 
```{r}
atlantic_arctic_spp <- read.csv("/home/kimberly.ledger/BeringSea_Arctic_eDNA/data/AtlanticArctic_SpeciesList.csv")

in_range <- spp_range %>%
  filter(in_range == "yes") %>%
  filter(!Species %in% atlantic_arctic_spp$Name)

#also_in_range <- range_na %>%
#  filter(in_range == "yes")

all_in_range <- in_range %>%
  #bind_rows(also_in_range) %>%
  bind_rows(not_Actinopteri_keepers)

## what species are in range but used a synonym? 
temp <- all_in_range %>%
  filter(Species %in% syn$Species)
temp 

#what names were used in the tax_df? 
syn %>%
  filter(Species %in% temp$Species)

#rename these in the sorted_tax_df
tax_df_temp <- tax_df_2 %>%
  mutate(species = ifelse(species == "Gadus ogac", "Gadus macrocephalus", species),
         taxon = ifelse(taxon == "Gadus ogac", "Gadus macrocephalus", taxon),
         species =  ifelse(species == "Embassichthys bathybius", "Microstomus bathybius", species),
         taxon = ifelse(taxon == "Embassichthys bathybius", "Microstomus bathybius", taxon),
         genus = ifelse(species == "Microstomus bathybius", "Microstomus", genus),
         species =  ifelse(species == "Ulcina olrikii", "Aspidophoroides olrikii", species),
         taxon = ifelse(taxon == "Ulcina olrikii", "Aspidophoroides olrikii", taxon),
         genus = ifelse(species == "Aspidophoroides olrikii", "Aspidophoroides", genus),
         #species = ifelse(species == "Polypera greeni", "Liparis greeni", species),
         #taxon = ifelse(taxon == "Polypera greeni", "Liparis greeni", taxon),
         #genus = ifelse(genus == "Polypera", "Liparis", genus),
         species = ifelse(species == "Sardinops melanostictus", "Sardinops sagax", species),
         taxon = ifelse(taxon == "Sardinops melanostictus", "Sardinops sagax", taxon))
tax_df_2 %>%
  filter(species == "Embassichthys bathybius")
tax_df_temp %>%
  filter(species == "Microstomus bathybius")
tax_df_3 <- tax_df_temp %>%
  filter(species %in% all_in_range$Species)
```

okay, now that out-of-range species have been removed, i need to remake the taxonomic table to reflect the more specific ids (hopefully)

## so now determine whether ASVs should be assigned to species genus, family, order, etc. 
```{r}
#remove taxon_level and taxon designations
tax_df_3 <- tax_df_3 %>%
  select(!taxonomic_level) %>%
  select(!taxon)

singletons <- tax_df_3 %>%
  select(qseqid) %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1)

# these are the seqs that now have only a single match
singleton_df <- singletons %>%
  left_join(tax_df_3) %>%
  select(-n) %>%
  mutate(taxonomic_level = "species") %>%
  mutate(taxon = species)

## Genus-level matches
# remove the singletons from the bigger df 
single_genus <- tax_df_3 %>%
  anti_join(singleton_df) %>% # at best, these should be genus-level matches
  group_by(qseqid, genus) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single genus
  select(-n) %>%
  left_join(., tax_df_3) %>%
  mutate(taxonomic_level = "genus") %>%
  mutate(taxon = genus)
  
## Family-level matches
single_family <- tax_df_3 %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  group_by(qseqid, family) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single family
  select(-n) %>%
  left_join(., tax_df_3) %>%
  mutate(taxonomic_level = "family") %>%
  mutate(taxon = family)


## Order-level matches
single_order <- tax_df_3 %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  group_by(qseqid, order) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single order
  select(-n) %>%
  left_join(., tax_df_3) %>%
  mutate(taxonomic_level = "order") %>%
  mutate(taxon = order)

## Class-level matches
single_class <- tax_df_3 %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  group_by(qseqid, class) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single class
  select(-n) %>%
  left_join(., tax_df_3) %>%
  mutate(taxonomic_level = "class") %>%
  mutate(taxon = class)

## Phylum-level matches
single_phylum <- tax_df_3 %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  anti_join(single_class) %>%
  group_by(qseqid, phylum) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single phylum
  select(-n) %>%
  left_join(., tax_df_3) %>%
  mutate(taxonomic_level = "phylum") %>%
  mutate(taxon = phylum)

# Kingdom level matches
single_kingdom <- tax_df_3 %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  anti_join(single_class) %>%
  group_by(qseqid, kingdom) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single kingdom
  select(-n) %>%
  left_join(., tax_df_3) %>%
  mutate(taxonomic_level = "kingdom") %>%
  mutate(taxon = kingdom)
```

## modify the singleton_df to include the right variable headers
```{r}
single_spp <- singleton_df
```

## recombine the full data set now that the appropriate level of taxonomy has been determined
```{r}
tax_df_4 <- bind_rows(single_kingdom, single_phylum, single_class, single_order, single_family, single_genus, single_spp)
```

just so i can pick up on this without having to rerun all the code, let me export is verison of the tax_df
```{r}
#write.csv(tax_df_4, "/home/kimberly.ledger/dbo_metabarcoding/outputs/tax_df_20240806.csv")
```
 
## output a list of all the Gadidae ASVs
```{r}
gadidae <- tax_df_4 %>%
  filter(family == "Gadidae") %>%
  select(qseqid) %>%
  unique()
```

## output a list of all the salmon ASVs
```{r}
salmon <- tax_df_4 %>%
  filter(genus == "Oncorhynchus") %>%
  select(qseqid) %>%
  unique()
```

## filter fasta to be just the gadids 
```{r}
# Load the necessary libraries
library(Biostrings)

# Define the input and output file paths
fasta_file <- "/home/kimberly.ledger/dbo_metabarcoding/data/NBS_SBS_DBO_20240806/filtered.seqs.fasta"
output_file <- "/home/kimberly.ledger/dbo_metabarcoding/outputs/gadidae.fasta"

# Read the FASTA file
fasta_seqs <- readDNAStringSet(fasta_file)

# Create a list of ASV sequence headers
asv_headers <- gadidae$qseqid  # 

# Filter the FASTA sequences based on the ASV headers
gadidae_seqs <- fasta_seqs[names(fasta_seqs) %in% asv_headers]

# Write the filtered sequences to a new FASTA file
writeXStringSet(gadidae_seqs, filepath = output_file)
```

## filter fasta to be just the gadids 
```{r}
# Load the necessary libraries
library(Biostrings)

# Define the input and output file paths
fasta_file <- "/home/kimberly.ledger/dbo_metabarcoding/data/NBS_SBS_DBO_20240806/filtered.seqs.fasta"
output_file <- "/home/kimberly.ledger/dbo_metabarcoding/outputs/oncorhynchus.fasta"

# Read the FASTA file
fasta_seqs <- readDNAStringSet(fasta_file)

# Create a list of ASV sequence headers
asv_headers <- salmon$qseqid  

# Filter the FASTA sequences based on the ASV headers
salmon_seqs <- fasta_seqs[names(fasta_seqs) %in% asv_headers]

# Write the filtered sequences to a new FASTA file
writeXStringSet(salmon_seqs, filepath = output_file)
```

in the command prompt (eDNA VM): 
(base) [kimberly.ledger@akc0ss-vu-134 outputs]$ blastn -query oncorhynchus.fasta -db ../../custom_dbs/oncorhynchus_db -out oncorhynchus_results.txt -perc_identity 98 -qcov_hsp_perc 98 -num_threads 10 -outfmt '6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sscinames staxids'

blastn -query gadidae.fasta -db ../../custom_dbs/gadidae_db -out gadidae_results.txt -perc_identity 98 -qcov_hsp_perc 98 -num_threads 10 -outfmt '6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sscinames staxids'

then see 1b_salmon_gadidae_taxonomic_assignment.Rmd

okay, so i cleaned up reference seq and blasted the gadidae and salmon reads against custom dbs. let me compare the assignments. 

```{r}
salmon_custom <- read.csv("/home/kimberly.ledger/dbo_metabarcoding/outputs/oncorhynchus_asv_id.csv") %>%
  select(!X)

salmon_original <- tax_df_4 %>%
  filter(genus == "Oncorhynchus") %>%
  select(qseqid, taxon, taxonomic_level) %>%
  unique()

salmon_join <- salmon_original %>%
  left_join(salmon_custom, by = "qseqid") %>%
  arrange(qseqid)
```


```{r}
gadidae_custom <- read.csv("/home/kimberly.ledger/dbo_metabarcoding/outputs/gadidae_asv_id.csv") %>%
  select(!X)

gadidae_original <- tax_df_4 %>%
  filter(family == "Gadidae") %>%
  select(qseqid, taxon, taxonomic_level) %>%
  unique()

gadidae_join <- gadidae_original %>%
  left_join(gadidae_custom, by = "qseqid") %>%
  arrange(qseqid)
```

for salmon and gadids, i'll use the custom reference DB results 
```{r}
gadid_salmon_tax_df <- gadidae_custom %>%
  bind_rows(salmon_custom)
```

now i still need to consider the rest of the IDs 

these are the single species (excluding gadidae and oncorhynchus)
```{r}
single_species <- tax_df_4 %>%
  filter(taxonomic_level == "species") %>%
  filter(family != "Gadidae") %>%
  filter(genus != "Oncorhynchus")
```

for a few groups, let's export assignments and see if there are taxonomic groups that make sense to use 
```{r}
sebastidae <- tax_df_4 %>%
  filter(taxonomic_level != "species")%>%
  filter(family == "Sebastidae") %>%
  filter(diff < 0.5)

#export to make custom groups 
write.csv(sebastidae, "/home/kimberly.ledger/dbo_metabarcoding/outputs/sebastidae_pre.csv", row.names = F)
```

```{r}
pleuronectidae <- tax_df_4 %>%
  filter(taxonomic_level != "species")%>%
  filter(family == "Pleuronectidae") %>%
  filter(diff < 0.5)

#export to make custom groups 
write.csv(pleuronectidae, "/home/kimberly.ledger/dbo_metabarcoding/outputs/pleuronectidae_pre.csv", row.names = F)
```


```{r}
cottidae <- tax_df_4 %>%
  filter(taxonomic_level != "species")%>%
  filter(family == "Cottidae") %>%
  filter(diff < 0.5)

#export to make custom groups 
write.csv(cottidae, "/home/kimberly.ledger/dbo_metabarcoding/outputs/cottidae_pre.csv", row.names = F)
```

okay, now just need to check out the missing tax ids 

```{r}
remaining <- tax_df_4 %>%
  filter(taxonomic_level != "species") %>%
  filter(family != "Gadidae") %>%
  filter(genus != "Oncorhynchus") %>%
  filter(family != "Sebastidae") %>%
  filter(family != "Pleuronectidae") %>%
  filter(family != "Cottidae") %>%
  filter(diff < 0.5)

#export for custom grouping 
write.csv(remaining, "/home/kimberly.ledger/dbo_metabarcoding/outputs/remaining_pre.csv", row.names = F)
```


read back in taxonomic groups
```{r}
sebastidae_post <- read.csv("/home/kimberly.ledger/dbo_metabarcoding/outputs/sebastidae_post.csv")
pleuronectidae_post <- read.csv("/home/kimberly.ledger/dbo_metabarcoding/outputs/pleuronectidae_post.csv")
cottidae_post <- read.csv("/home/kimberly.ledger/dbo_metabarcoding/outputs/cottidae_post.csv")
remaining_post <- read.csv("/home/kimberly.ledger/dbo_metabarcoding/outputs/remaining_post.csv")
```


add the single species and the families back to the final tax df 
```{r}
tax_df_5 <- single_species %>%
  bind_rows(sebastidae_post) %>%
  bind_rows(pleuronectidae_post) %>%
  bind_rows(cottidae_post) %>%
  bind_rows(remaining_post) %>%
  bind_rows(gadid_salmon_tax_df)
```

double check i didn't forget any ASVs in this process
```{r}
filtered_unique <- tax_df_4 %>%
  select(qseqid) %>%
  unique() 

# Check if all elements in table1$qseqid are in table2$qseq
match <- all(filtered_unique$qseqid %in% tax_df_5$qseqid)

# Print result
if (match) {
  print("All rows in 'qseqid' column of table 1 match the rows in 'qseqid' column of table 2.")
} else {
  print("Not all rows in 'qseqid' column of table 1 match the rows in 'qseqid' column of table 2.")
}

# missing <- filtered_unique %>%
#   anti_join(tax_df_5)
# 
# tax_df_2 %>%
#   filter(qseqid %in% missing$qseqid)
# 
# tax_df_4 %>%
#   filter(qseqid %in% missing$qseqid)
# 
# ## top matches are not in range and diff to matches in range was over 0.5%. i'll add these back in manually. 
# 
# missing$taxon <- c("Stichaeidae 2",  "Hexagrammos", "Ammodytes")
# missing$taxonomic_level <-c("family", "genus", "genus")
# 
# tax_df_6 <- tax_df_5 %>%
#   bind_rows(missing)
# 
# # Check if all elements in table1$qseqid are in table2$qseq
# match <- all(filtered_unique$qseqid %in% tax_df_6$qseqid)
# 
# # Print result
# if (match) {
#   print("All rows in 'qseqid' column of table 1 match the rows in 'qseqid' column of table 2.")
# } else {
#   print("Not all rows in 'qseqid' column of table 1 match the rows in 'qseqid' column of table 2.")
# }
```

## Create output taxonomy data frames
```{r}
# and then collapse that down to just a single taxon per ASV
# can i collapse while keeping the taxonomy for the taxon....? 
tax_df_6 <- tax_df_5 %>%
  mutate(species = ifelse(taxonomic_level == "family", NA, species),
         genus = ifelse(taxonomic_level == "family", NA, genus),
         species = ifelse(taxonomic_level == "genus", NA, species)) %>%
  select(qseqid, taxon, taxonomic_level, species, genus, family, order, class, phylum, kingdom) %>%
  unique()
```

```{r}
write.csv(tax_df_5, "/home/kimberly.ledger/dbo_metabarcoding/outputs/taxonomy_20240806_uncollapsed.csv")
write.csv(tax_df_6, "/home/kimberly.ledger/dbo_metabarcoding/outputs/taxonomy_20240806_collapsed.csv")
```

how many unique taxa? 
```{r}
length(unique(tax_df_6$taxon))
unique(tax_df_6$taxon)
```

output a table that has the taxon descriptions (i.e. what species make up a genus/family level assignment)
```{r}
taxon_groups <- tax_df_5 %>%
  select(!qseqid) %>%
  select(!seq_percID) %>%
  select(!top_perc) %>%
  select(!diff) %>%
  unique() 

write.csv(taxon_groups, "/home/kimberly.ledger/dbo_metabarcoding/outputs/taxon_groups.csv")
```
